File: server/modules/workspace/files.ts
import { open, readdir, stat, writeFile } from "node:fs/promises";
import { join, relative, sep } from "node:path";
import { assertPathWithinRoot } from "./path-guard.js";

export const MAX_FILE_PREVIEW_BYTES = 512 * 1024;

export class FileReadError extends Error {
  constructor(
    message: string,
    readonly status: 400 | 404 | 413,
  ) {
    super(message);
    this.name = "FileReadError";
  }
}

const NOISE_DIRS = new Set([
  "node_modules",
  ".git",
  ".next",
  "dist",
  ".turbo",
  ".cache",
  "__pycache__",
  ".venv",
  ".DS_Store",
]);

export async function listFilesForWorktree(worktreePath: string): Promise<string[]> {
  const paths: string[] = [];

  async function walk(dir: string): Promise<void> {
    let entries;
    try {
      entries = await readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (NOISE_DIRS.has(entry.name)) continue;
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(fullPath);
      } else {
        const rel = relative(worktreePath, fullPath);
        paths.push(rel.split(sep).join("/"));
      }
    }
  }

  await walk(worktreePath);
  return paths;
}

export async function readFileForWorktree(
  worktreePath: string,
  relativePath: string,
): Promise<{ path: string; content: string; truncated: boolean }> {
  const normalized = relativePath.replace(/\\/g, "/").replace(/^\/+/, "");
  if (!normalized || normalized.endsWith("/")) {
    throw new FileReadError("Invalid file path", 400);
  }

  const absolutePath = assertPathWithinRoot(worktreePath, normalized);

  let fileStat;
  try {
    fileStat = await stat(absolutePath);
  } catch {
    throw new FileReadError("File not found", 404);
  }

  if (!fileStat.isFile()) {
    throw new FileReadError("Not a file", 400);
  }

  const truncated = fileStat.size > MAX_FILE_PREVIEW_BYTES;
  const readLength = truncated ? MAX_FILE_PREVIEW_BYTES : fileStat.size;

  const handle = await open(absolutePath, "r");
  try {
    const buf = Buffer.alloc(readLength);
    const { bytesRead } = await handle.read(buf, 0, readLength, 0);
    const slice = buf.subarray(0, bytesRead);
    if (slice.includes(0)) {
      throw new FileReadError("Binary file cannot be previewed", 400);
    }
    return {
      path: normalized,
      content: slice.toString("utf8"),
      truncated,
    };
  } finally {
    await handle.close();
  }
}

export async function writeFileForWorktree(
  worktreePath: string,
  relativePath: string,
  content: string,
): Promise<void> {
  const normalized = relativePath.replace(/\\/g, "/").replace(/^\/+/, "");
  if (!normalized || normalized.endsWith("/")) {
    throw new FileReadError("Invalid file path", 400);
  }

  const absolutePath = assertPathWithinRoot(worktreePath, normalized);

  let fileStat;
  try {
    fileStat = await stat(absolutePath);
  } catch {
    throw new FileReadError("File not found", 400);
  }

  if (!fileStat.isFile()) {
    throw new FileReadError("Not a file", 400);
  }

  try {
    await writeFile(absolutePath, content, "utf-8");
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : "Write failed";
    throw new FileReadError(`Failed to write file: ${msg}`, 400);
  }
}

export async function searchFilesForWorktree(
  worktreePath: string,
  query: string,
  limit: number,
): Promise<string[]> {
  const allPaths = await listFilesForWorktree(worktreePath);
  const q = query.toLowerCase();

  const scored: Array<{ path: string; score: number }> = [];
  for (const p of allPaths) {
    const lower = p.toLowerCase();
    const filename = lower.split("/").at(-1) ?? lower;
    let score = 0;
    if (filename.startsWith(q)) score = 3;
    else if (filename.includes(q)) score = 2;
    else if (lower.includes(q)) score = 1;
    if (score > 0) scored.push({ path: p, score });
  }

  return scored
    .sort((a, b) => b.score - a.score)
    .slice(0, limit)
    .map((s) => s.path);
}

File: server-rs/src/workspace/files.rs
use crate::error::{file_read_error, AppError, AppResult};
use serde::Serialize;
use std::fs;
use std::io::Read;
use std::path::Path;

use super::path_guard::assert_path_within_root;

pub const MAX_FILE_PREVIEW_BYTES: usize = 512 * 1024;

const NOISE_DIRS: &[&str] = &[
    "node_modules",
    ".git",
    ".next",
    "dist",
    ".turbo",
    ".cache",
    "__pycache__",
    ".venv",
    ".DS_Store",
];

#[derive(Serialize)]
pub struct FilePreview {
    pub path: String,
    pub content: String,
    pub truncated: bool,
}

fn normalize_relative_path(relative_path: &str) -> Result<String, AppError> {
    let normalized = relative_path.replace('\\', "/").trim_start_matches('/').to_string();
    if normalized.is_empty() || normalized.ends_with('/') {
        return Err(file_read_error("Invalid file path", 400));
    }
    Ok(normalized)
}

pub fn list_files_for_worktree(worktree_path: &str) -> AppResult<Vec<String>> {
    let mut paths = Vec::new();
    walk_dir(Path::new(worktree_path), Path::new(worktree_path), &mut paths)?;
    Ok(paths)
}

fn walk_dir(worktree_path: &Path, dir: &Path, paths: &mut Vec<String>) -> AppResult<()> {
    let entries = match fs::read_dir(dir) {
        Ok(entries) => entries,
        Err(_) => return Ok(()),
    };

    for entry in entries.flatten() {
        let name = entry.file_name();
        let name = name.to_string_lossy();
        if NOISE_DIRS.contains(&name.as_ref()) {
            continue;
        }
        let full_path = entry.path();
        if full_path.is_dir() {
            walk_dir(worktree_path, &full_path, paths)?;
        } else if full_path.is_file() {
            let rel = full_path
                .strip_prefix(worktree_path)
                .unwrap_or(&full_path)
                .to_string_lossy()
                .replace('\\', "/");
            paths.push(rel);
        }
    }
    Ok(())
}

pub fn read_file_for_worktree(worktree_path: &str, relative_path: &str) -> AppResult<FilePreview> {
    let normalized = normalize_relative_path(relative_path)?;
    let absolute_path = assert_path_within_root(worktree_path, &normalized)
        .map_err(|e| file_read_error(e, 400))?;

    let metadata = fs::metadata(&absolute_path).map_err(|_| file_read_error("File not found", 404))?;
    if !metadata.is_file() {
        return Err(file_read_error("Not a file", 400));
    }

    let truncated = metadata.len() as usize > MAX_FILE_PREVIEW_BYTES;
    let read_length = if truncated {
        MAX_FILE_PREVIEW_BYTES
    } else {
        metadata.len() as usize
    };

    let mut file = fs::File::open(&absolute_path)
        .map_err(|_| file_read_error("File not found", 404))?;
    let mut buf = vec![0u8; read_length];
    let bytes_read = file
        .read(&mut buf)
        .map_err(|e| file_read_error(format!("Read failed: {e}"), 400))?;
    buf.truncate(bytes_read);

    if buf.contains(&0) {
        return Err(file_read_error("Binary file cannot be previewed", 400));
    }

    let content = String::from_utf8(buf).map_err(|_| file_read_error("Invalid UTF-8", 400))?;
    Ok(FilePreview {
        path: normalized,
        content,
        truncated,
    })
}

pub fn write_file_for_worktree(
    worktree_path: &str,
    relative_path: &str,
    content: &str,
) -> AppResult<()> {
    let normalized = normalize_relative_path(relative_path)?;
    let absolute_path = assert_path_within_root(worktree_path, &normalized)
        .map_err(|e| file_read_error(e, 400))?;

    let metadata = fs::metadata(&absolute_path).map_err(|_| file_read_error("File not found", 400))?;
    if !metadata.is_file() {
        return Err(file_read_error("Not a file", 400));
    }

    fs::write(&absolute_path, content).map_err(|e| {
        file_read_error(format!("Failed to write file: {e}"), 400)
    })
}
